parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java

```diff
@@ -1132,6 +1138,109 @@ public ColumnChunkPageReadStore readFilteredRowGroup(int blockIndex, RowRanges r
     return internalReadFilteredRowGroup(block, rowRanges, getColumnIndexStore(blockIndex));
   }
 
+  /**
+   * Read data in all parts via either vectored IO or serial IO.
+   * @param allParts all parts to be read.
+   * @param builder used to build chunk list to read the pages for the different columns.
+   * @throws IOException any IOE.
+   */
+  private void readAllPartsVectoredOrNormal(List<ConsecutivePartList> allParts, ChunkListBuilder builder)
+      throws IOException {
+
+    if (shouldUseVectoredIO(allParts)) {
+      try {
+        readVectored(allParts, builder);
+        return;
+      } catch (IOException | IllegalArgumentException | UnsupportedOperationException e) {
+        // possible failure modes.
+        LOG.warn("readVectored() failed; falling back to normal IO", e);
+      }
+    }
+    for (ConsecutivePartList consecutiveChunks : allParts) {
+      consecutiveChunks.readAll(f, builder);
+    }
+  }
+
+  /**
+   * Should the read use vectored IO?
+   * <p>
+   * This returns true if all necessary conditions are met:
+   * <ol>
+   *   <li> The option is enabled</li>
+   *   <li> The Hadoop version supports vectored IO</li>
+   *   <li> The part lengths are all valid for vectored IO</li>
+   *   <li> The stream implementation explicitly supports the API; for other streams the classic
+   *         API is always used.</li>
+   *   <li> The allocator is not direct. This is to avoid HADOOP-19101 surfacing.
+   * </ol>
+   * @param allParts all parts to read.
+   * @return true or false.
+   */
+  private boolean shouldUseVectoredIO(final List<ConsecutivePartList> allParts) {
+    return options.useHadoopVectoredIO()
+        && !options.getAllocator().isDirect()
+        && f.readVectoredAvailable()
+        && arePartsValidForVectoredIO(allParts);
+  }
+
+  /**
+   * Validated the parts for vectored IO.
+   * Vectored IO doesn't support reading ranges of size greater than
+   * Integer.MAX_VALUE.
+   * @param allParts all parts to read.
+   * @return true or false.
+   */
+  private boolean arePartsValidForVectoredIO(List<ConsecutivePartList> allParts) {
+    for (ConsecutivePartList consecutivePart : allParts) {
```

Comment:
I thought we decided to add a check for overlapping ranges explicitly in Parquet. where is that?

Reply:
not yet ready; i want to make sure the hadoop-side validation does what I want and then will copy and paste.

One thing there is to decide how best to deal with negative offsets. EOF vs IllegalArgument?

Reply:
+1 for IllegalArgument on my side. IMO, fallback for overlapping ranges is an overkill. We can simply throw and let the user to turn off vectored io if something is wrong.

Reply:
missed this, will tweak